ngraph/core/src/op/gather.cpp

```diff
@@ -128,6 +128,166 @@ shared_ptr<Node> op::v1::Gather::clone_with_new_inputs(const OutputVector& new_a
     return make_shared<v1::Gather>(new_args.at(PARAMS), new_args.at(INDICES), new_args.at(AXIS));
 }
 
+NGRAPH_RTTI_DEFINITION(op::v7::Gather, "Gather", 7);
+static constexpr int64_t AXIS_NOT_SET_VALUE = std::numeric_limits<int64_t>::max();
+
+op::v7::Gather::Gather(const Output<Node>& data,
+                       const Output<Node>& indices,
+                       const Output<Node>& axis,
+                       const int64_t batch_dims)
+    : Op({data, indices, axis})
+    , m_batch_dims(batch_dims)
+{
+    constructor_validate_and_infer_types();
+}
+
+bool ngraph::op::v7::Gather::visit_attributes(AttributeVisitor& visitor)
+{
+    NGRAPH_OP_SCOPE(v7_Gather_visit_attributes);
+    visitor.on_attribute("batch_dims", m_batch_dims);
+    return true;
+}
+
+void op::v7::Gather::validate_and_infer_types()
+{
+    NGRAPH_OP_SCOPE(v7_Gather_validate_and_infer_types);
+    const auto& data_type = get_input_element_type(0);
+    const auto& indices_type = get_input_element_type(1);
+
+    NODE_VALIDATION_CHECK(this,
+                          indices_type == element::Type_t::i32 ||
+                              indices_type == element::Type_t::i64,
+                          "indices must be of int32 or int64 type. But instead got: ",
+                          indices_type);
+
+    const auto& data_pshape = get_input_partial_shape(0);
+    const auto& indices_pshape = get_input_partial_shape(1);
+    const auto& axis_pshape = get_input_partial_shape(2);
+    auto data_rank = data_pshape.rank();
+    auto indices_rank = indices_pshape.rank();
+    auto axis_rank = axis_pshape.rank();
+
+    if (axis_rank.is_static() && axis_pshape.is_static())
+    {
+        const auto axis_is_scalar = axis_rank.get_length() == 0;
+        const auto axis_has_one_elem =
+            axis_rank.get_length() == 1 && axis_pshape[0].get_length() == 1;
+        NODE_VALIDATION_CHECK(
+            this,
+            axis_is_scalar || axis_has_one_elem,
+            "Axes input must be scalar or have 1 element. But instead got axis_shape = ",
+            axis_pshape);
+    }
+
+    const auto& axes_constant = get_constant_from_source(input_value(2));
+    if (!axes_constant || axes_constant->cast_vector<int64_t>().empty())
+    {
+        set_output_type(0, data_type, PartialShape::dynamic());
```

Comment:
Why can calculate the rank of the output tensor if the ranks of input tensor are known. So the fully dynamic output tensor is not the best option.

Reply:
Moreover, the shape of the first batch_dims can be copied from the corresponding input.

Please, add several unit tests to cover these scenarious

Reply:
I also thought about this but there are several difficulties. For out_rank is bound to values of `batch_dims`. 
>out_rank = data_rank + indices_rank  - batch_dims - 1

We have some restrictions for `batch_dims` - it cannot be greater than `axis`. Also if `batch_dims` for some reason will arrive negative (if in future we will support reading through TF_IMPORTER w/o MO) we will not be able to resolve it without axis.

We can do optimistic assumptions and skip the first check (if axis is not set) but in that case code will be less consistent: either we will be using in one place getter for `batch_dims` in another without getters or manually convert `batch_dims` to positive in every single place we need it.

So I don't know what is the optimal solution here. Where the shape will be calculated when `axis` will arrive? Here in nGraph or somewhere in plugins? If here, then as soon as it will arrive we will calculate actual out_rank from scratch in `validate_and_infer_types`. Let's discuss how to deal with this personally.

Reply:
Changed according to your suggestion. Added unit-tests for that cases as well